sonarts-core/src/rules/noUnusedArrayRule.ts

```diff
@@ -0,0 +1,195 @@
+
+export class Rule extends tslint.Rules.TypedRule {
+  public static metadata: SonarRuleMetaData = {
+    ruleName: "no-unused-array",
+    description: "Array contents should be used",
+    optionsDescription: "",
+    options: null,
+    rspecKey: "RSPEC-4030",
+    type: "maintainability",
+    typescriptOnly: false,
+  };
+
+  private static writeArrayPatterns: ((statement: ts.ExpressionStatement, usage: Usage) => boolean)[] = [
+    Rule.isElementWrite,
+    Rule.isVariableWrite,
+    Rule.isWritingMethodCall,
+  ];
+
+  public static MESSAGE = "Either use this array's contents or remove the array.";
+
+  public applyWithProgram(sourceFile: ts.SourceFile, program: ts.Program): tslint.RuleFailure[] {
+    const symbols = SymbolTableBuilder.build(sourceFile, program);
+
+    // walker is created to only save issues
+    const walker = new tslint.RuleWalker(sourceFile, this.getOptions());
+    const symbolDeclarations = new Map();
+
+    symbols
+      .getSymbols()
+      // get only symbols storing arrays
+      .filter(symbol => Rule.hasArrayType(symbol, program.getTypeChecker(), symbols))
+      // filter out unused symbols
+      .filter(symbol => symbols.allUsages(symbol).length > 1)
+      // filter out symbols without declaration
+      .filter(symbol => {
+        const declaration = Rule.getDeclaration(symbol, symbols);
+        if (declaration) {
+          symbolDeclarations.set(symbol, declaration);
+        }
+        return declaration;
+      })
+      // filter out parameters and exported/imported symbols
+      .filter(
+        symbol =>
+          !firstAncestor(symbolDeclarations.get(symbol), [ts.SyntaxKind.Parameter, ts.SyntaxKind.ImportDeclaration]) &&
+          !Rule.isExported(symbolDeclarations.get(symbol)),
+      )
+      // keep only symbols initialized to array literal or not initialized at all
+      .filter(symbol => {
+        const varDeclaration = firstAncestor(symbolDeclarations.get(symbol), [
+          ts.SyntaxKind.VariableDeclaration,
+        ]) as ts.VariableDeclaration;
+        if (varDeclaration) {
+          return !varDeclaration.initializer || Rule.isArrayLiteral(varDeclaration.initializer);
+        }
+        return true;
+      })
+      // filter out symbols with at least one read usage
+      .filter(symbol => !symbols.allUsages(symbol).some(usage => Rule.isReadUsage(usage)))
+      // raise issue
+      .forEach(symbol => walker.addFailureAtNode(symbolDeclarations.get(symbol), Rule.MESSAGE));
+
+    return walker.getFailures();
+  }
+
+  private static isExported(declaration: ts.Node): boolean {
+    const varStatement = firstAncestor(declaration, [ts.SyntaxKind.VariableStatement]);
+    return !!varStatement && is(varStatement.getFirstToken(), ts.SyntaxKind.ExportKeyword);
+  }
+
+  private static isReadUsage(usage: Usage): boolean {
+    if (usage.is(UsageFlag.DECLARATION)) {
+      return false;
+    }
+
+    const expressionStatement = firstAncestor(usage.node, [
```

Comment:
@vilchik-elena do you know if we can change this formatting to something that is not this ugly? This should just be `firstAncestor(usage.node, [ ts.SyntaxKind.ExpressionStatement ])` in one line.